File: 2025/SoyoungJung/week11/2025-08-17-Longest-Palindromic-Substring.ts
function longestPalindrome(s: string): string {
  if (s.length < 2) return s;

  let start = 0;
  let maxLen = 1;

  const expandFromCenter = (left: number, right: number): void => {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      if (right - left + 1 > maxLen) {
        maxLen = right - left + 1;
        start = left;
      }
      left--;
      right++;
    }
  };

  for (let i = 0; i < s.length; i++) {
    expandFromCenter(i, i);
    expandFromCenter(i, i + 1);
  }

  return s.substring(start, start + maxLen);
}

File: 2025/jimmycho/week11/2025-08-17-Longest-Palindromic-Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[False] * len(s) for _ in range(len(s))]
        for i in range(len(s)):
            dp[i][i] = True

        res = (0, 0, 0)
        for end in range(1, len(s)):
            for start in range(end):
                if s[start] == s[end]:
                    if end - start == 1:
                        dp[end][start] = True
                    else:
                        dp[end][start] = dp[end - 1][start + 1]
                    if dp[end][start] and end - start > res[0]:
                        res = (end - start, start, end)
        return s[res[1]: res[2] + 1]